group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -806,7 +809,29 @@ public void validateOffsetCommit(
         if (memberEpoch < 0 && members().isEmpty()) return;
 
         final ConsumerGroupMember member = getOrMaybeCreateMember(memberId, false);
-        validateMemberEpoch(memberEpoch, member.memberEpoch());
+        if (member.useClassicProtocol()) {
+            validateMemberInstanceId(member, groupInstanceId);
+
+            try {
+                validateMemberEpoch(memberEpoch, member.memberEpoch());
+            } catch (StaleMemberEpochException ex) {
+                // StaleMemberEpochException is not supported in the classic protocol. We throw
+                // IllegalGenerationException instead for compatibility.
+                throw new IllegalGenerationException(String.format("Invalid offset commit because the "
+                    + "received generation id %d does not match the expected member epoch %d.",
+                    memberEpoch, member.memberEpoch()));
+            }
+
+            if (member.memberEpoch() < groupEpoch() ||
+                member.state() == MemberState.UNREVOKED_PARTITIONS ||
+                (member.state() == MemberState.UNRELEASED_PARTITIONS && !waitingOnUnreleasedPartition(member))) {
+                throw new RebalanceInProgressException(String.format("Invalid offset commit because" +
+                    " a new rebalance has been triggered in group %s and member %s should rejoin to catch up.",
+                    groupId(), member.memberId()));
+            }
```

Comment:
I am not sure about this logic here. Let's imagine that a member must rebalance and it got the signal via the heartbeat request. The first thing that it will do is committing its offsets. With this logic, the commits will be rejected. Therefore, I wonder if we should remove it and rely on the other APIs to trigger rebalances. What do you think?

Reply:
Got it, now I know why the classic consumer offset commit validation doesn't check PREPARING_REBALANCE.